usr.bin/xstr/xstr.1
.\"	$NetBSD: xstr.1,v 1.19.6.1 2019/06/10 22:10:27 christos Exp $
.\"
.\" Copyright (c) 1980, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)xstr.1	8.2 (Berkeley) 12/30/93
.\"
.Dd July 23, 2004
.Dt XSTR 1
.Os
.Sh NAME
.Nm xstr
.Nd "extract strings from C programs to implement shared strings"
.Sh SYNOPSIS
.Nm
.Op Fl cv
.Op Fl l Ar array
.Op Fl
.Op Ar
.Sh DESCRIPTION
.Nm
maintains a file
.Pa strings
into which strings in component parts of a large program are hashed.
These strings are replaced with references to this common area.
This serves to implement shared constant strings, most useful if they
are also read-only.
.Pp
Available options:
.Bl -tag -width XXlXarrayXX
.It Fl
.Nm
reads from the standard input.
.It Fl c
.Nm
will extract the strings from the C source
.Ar file
or the standard input
.Pq Fl ,
replacing
string references by expressions of the form (&xstr[number])
for some number.
An appropriate declaration of
.Nm
is prepended to the file.
The resulting C text is placed in the file
.Pa x.c ,
to then be compiled.
The strings from this file are placed in the
.Pa strings
data base if they are not there already.
Repeated strings and strings which are suffixes of existing strings
do not cause changes to the data base.
.It Fl l Ar array
Specify the named array in program references to abstracted
strings.
The default array name is xstr.
.It Fl v
Be verbose.
.El
.Pp
After all components of a large program have been compiled, a file
.Pa xs.c
declaring the common
.Nm
space can be created by a command of the form:
.Pp
.Dl $ xstr
.Pp
The file
.Pa xs.c
should then be compiled and loaded with the rest
of the program.
If possible, the array can be made read-only (shared) saving
space and swap overhead.
.Pp
.Nm
can also be used on a single file.
The following command creates files
.Pa x.c
and
.Pa xs.c
as before, without using or affecting any
.Pa strings
file in the same directory:
.Pp
.Dl $ xstr name
.Pp
It may be useful to run
.Nm
after the C preprocessor if any macro definitions yield strings
or if there is conditional code which contains strings
which may not, in fact, be needed.
An appropriate command sequence for running
.Nm
after the C preprocessor is:
.Bd -literal -offset indent
$ cc \-E name.c | xstr \-c \-
$ cc \-c x.c
$ mv x.o name.o
.Ed
.Pp
.Nm
does not touch the file
.Pa strings
unless new items are added, thus
.Xr make 1
can avoid remaking
.Pa xs.o
unless truly necessary.
.Sh FILES
.Bl -tag -width /tmp/xsxx* -compact
.It Pa strings
Data base of strings
.It Pa x.c
Massaged C source
.It Pa xs.c
C source for definition of array `xstr'
.It Pa /tmp/xs*
Temp file when `xstr name' doesn't touch
.Pa strings
.El
.Sh SEE ALSO
.Xr mkstr 1
.Sh HISTORY
The
.Nm
command appeared in
.Bx 3.0 .
.Sh BUGS
If a string is a suffix of another string in the data base,
but the shorter string is seen first by
.Nm
both strings will be placed in the data base, when just
placing the longer one there will do.
.Pp
.Nm
does not parse the file properly so it does not know not to process:
.Bd -literal
	char var[] = "const";
.Ed
into:
.Bd -literal
	char var[] = (&xstr[N]);
.Ed
.Pp
These must be changed manually into an appropriate initialization for
the string, or use the following ugly hack.
.Pp
Also,
.Nm
cannot initialize structures and unions that contain strings.
Those can be fixed by changing from:
.Bd -literal
	struct foo {
		int i;
		char buf[10];
	} = {
		1, "foo"
	};
.Ed
to:
.Bd -literal
	struct foo {
		int i;
		char buf[10];
	} = {
		1, { 'f', 'o', 'o', '\e0' }
	};
.Ed
.Pp
The real problem in both cases above is that the compiler knows the size
of the literal constant so that it can perform the initialization required,
but when
.Nm
changes the literal string to a pointer reference, the size information is
lost.
It would require a real parser to do this right, so the obvious solution is
to fix the program manually to compile, or even better rely on the compiler
and the linker to merge strings appropriately.
.Pp
Finally,
.Nm
is not very useful these days because most of the string merging is done
automatically by the compiler and the linker, provided that the strings
are identical and read-only.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2405.2.1 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 8.0 to NetBSD 9.0:
	byacc: Update to 20170430. [christos 20170605]
	vax: Add support for VAXstation 4000 TURBOchannel. [flxd 20170609]
	wsbell(4): Added console bell support for all speaker devices, not
		only those attached at pcppi. [nat 20170612]
	bind: Import version 9.10.5-P1. [christos 20170615]
	dts: Import dts files from Linux 4.11.5. [jmcneill 20170615]
	amd64: Support 16TB of PA, and 32TB of kernel VA on the architecture
		side. [maxv 20170617]
	expat: Import 2.2.1 (security fixes). [christos 20170617]
	evbarm: Add support for Allwinner H3 SoC. [jmcneill 20170628]
	bind: Import version 9.10.5-P2. [christos 20170630]
	sunxiemac(4): Add support for Allwinner Gigabit Ethernet (EMAC).
		[jmcneill 20170701]
	fsdb(8): Add "saveblks" command. [mrg 20170804]
	wbsio(4): Add support for Winbond W83627DHG-P, W83627SF, W83627UHG,
		W83667HGB, W83687THF, W83697UG,	Nuvoton NCT5104D, NCT6775,
		NCT6779, NCT6791, NCT6792 and NCT6793. [msaitoh 20170707]
	evbarm: Add support for Allwinner A83T SoC. [jmcneill 20170707]
	MAKEDEV(8): Regenerate, documenting lua, dk, plcom, wmcom, hdmicec, mfi,
		nvme, nvme ns, tpm, and dtrace devices. [sevan 20170710]
	wbsio(4): Add support for NCT6795. [msaitoh 20170711]
	lm(4): Add support for NCT5174D, NCT6775F, NCT6779D and NCT679[1235]D.
		[msaitoh 20170711]
	evbarm: Add support for NVIDIA Tegra X1 SoC. [jmcneill 20170720]
	evbarm: Add support for Allwinner A64 SoC. [jmcneill 20170723]
	gzip(1): Add SIGINFO support. [mrg 20170804]
	i386: Remove vm86. [maxv 20170812]
	itesio(4): Add support IT8628E. [hauke 20170814]
	ichsmb(4): Add support for Intel C3000. [msaitoh 20170817]
	ismt(4): Add support for Intel C3000. [msaitoh 20170817]
	powerpc: Bump MAXTSIZ to 128MB for OEA based systems. [sevan 20170818]
	GMP/MPFR/MPC: Update to 6.1.2, 3.1.5 and 1.0.3. [mrg 20170821]
	ixg(4), ixv(4): Add C3000 support. Add bypass function support for
		bypass adapters. [msaitoh 20170830]
	file(1): Upgraded to 5.32. [christos 20170908]
	tcpdump(8): Import 4.9.2. [christos 20170908]
	veriexecgen(8): Drop support for MD5, SHA1, RMD160. [sevan 20170909]
	itesio(4): Add support IT8728GF and IT877[12]E [hauke 20170814]
	evbarm: Add support for Allwinner A13 and R8 SoCs. [jmcneill 20170825]
	veriexec(4): Drop support for VERIFIED_EXEC_FP_MD5,
		VERIFIED_EXEC_FP_SHA1, and VERIFIED_EXEC_FP_RMD160 options.
		[sevan 20170913]
	acpi(4): Updated ACPICA to 20170831. [christos 20170915]
	dtrace: Install dtruss, execsnoop, opensnoop, and procsystime from the
		DTrace Toolkit if DTrace support is enabled. [sevan 20170917]
	dhcpcd(8): Import dhcpcd-7.0.0-rc3. [roy 20171007]
	ahcisata(4), siisata(4): Add support for NCQ. [jdolecek 20171007]
	mvsata(4): Add support for NCQ, and enable ATAPI support.
		[jdolecek 20171007]
	OpenSSH: Imported 7.6. [christos 20171007]
	amd64: Add support for Kernel ASLR [maxv 20171010]
	evbarm: Add support for Allwinner H5 SoC. [jmcneill 20171011]
	tmux(1): Imported 2.6. [christos 20171012]
	amd64: Add support for SMAP. [maxv 20171017]
	amd64: Add support for USER_LDT. [maxv 20171021]
	tzdata Updated to 2017c. [kre 20171024]
	libc: Update to tzcode2017c. [christos 20171024]
	bwfm(4): Add driver for Broadcom BCM43xxx "FullMAC" USB wireless
		adapters. [jmcneill 20171025]
	kernhist(9): Reworked the export-via-sysctl code to use uintptr_t
		arguments, PR kern/52639.  Requires contemporaneous update
		of vmstat(1). [pgoyette 20171028]
	evbarm: Add support for NextThing GR8 SoC. [jmcneill 20171106]
	acpi(4): Updated ACPICA to 20171110. [christos 20171111]
	sunxinand(4): Add driver for Allwinner NAND Flash Controller.
		[jmcneill 20171113]
	gcc(1): Import GCC 5.5.  [mrg 20171113]
	libtre: Update to the latest git source. [rin 20171117]
	qemufwcfg(4): Add driver for QEMU Firmware Configuration device.
		[jmcneill 20171125]
	gdb(1): Updated to 8.0.1.  [christos 20171128]
	dhcpcd(8): Import dhcpcd-7.0.0-rc4. [roy 20171206]
	ihidev(4), ims(4): Added drivers for i2c HID mice. [bouyer 20171210]
	sparc: Fix time goes backwards. [mrg 20180112]
	lua: Applied 6th bugfix to Lua 5.3.4 from lua.org/bugs.html.
		[mbalmer 20171213]
	sstk: Remove the "stack section size change" syscall [kamil 20171219]
	sbrk(2): Remove the "data section size change" syscall [kamil 20171219]
	ovadvise: Remove the old vadvise syscall header [kamil 20171219]
	dhcpcd(8): Import dhcpcd-7.0.0 [roy 20180101]
	mdnsd(8), dns-sd(1), libdns_sd: import mDNSResponder 878.1
		[christos 20180114]
	cpuctl(8): Add cpuctl identify support for sparc and sparc64.
		[mrg 20180116]
	macppc: Enable Veriexec support in the GENERIC kernel [sevan 20180117]
	raid(4): Add support for 32-bit on 64-bit kernel.  [mrg 20180117]
	zoneinfo: Import tzdata2018c (skipping 2018a & 2018b). [kre 20180124]
	libc: Update to tzcode2018c. [christos 20180125]
	i386: Add support for SMAP. [maxv 20180128]
	dhcpcd(8): Import dhcpcd-7.0.1 [roy 20180129]
	gcc(1): Import GCC 6.4.  [mrg 20180201]
	openldap: Import 2.4.45. [christos 20180205]

usr.bin/xstr/xstr.c
/*	$NetBSD: xstr.c,v 1.26 2016/03/11 18:39:03 christos Exp $	*/

/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)xstr.c	8.1 (Berkeley) 6/9/93";
#else
__RCSID("$NetBSD: xstr.c,v 1.26 2016/03/11 18:39:03 christos Exp $");
#endif
#endif /* not lint */

#include <sys/param.h>
#include <signal.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#include <err.h>
#include "pathnames.h"

/*
 * xstr - extract and hash strings in a C program
 *
 * Bill Joy UCB
 * November, 1978
 */

static off_t	hashit(const char *, int);
__dead static void	onintr(int);
static off_t	yankstr(char **);
static int	octdigit(char);
static void	inithash(void);
static int	fgetNUL(char *, int, FILE *);
static int	xgetc(FILE *);
static void	flushsh(void);
static void	found(int, off_t, const char *);
static void	prstr(const char *);
static void	xsdotc(void);
static char	lastchr(const char *);
static int	istail(const char *, const char *);
static void	process(const char *);
__dead static void	usage(void);

static off_t	tellpt;
static off_t	mesgpt;
static char	stringtmpfile[MAXPATHLEN];
static const char *strings =	"strings";
static const char *array =	0;
static int	cflg;
static int	vflg;
static int	readstd;
static char	linebuf[8192];

#define	BUCKETS	128

static struct	hash {
	off_t	hpt;
	char	*hstr;
	struct	hash *hnext;
	short	hnew;
} bucket[BUCKETS];

int
main(int argc, char *argv[])
{
	int c;

	while ((c = getopt(argc, argv, "-cvl:")) != -1)
		switch (c) {
		case '-':
			readstd++;
			break;
		case 'c':
			cflg++;
			break;
		case 'v':
			vflg++;
			break;
		case 'l':
			array = optarg;
			break;
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (array == 0)
		array = "xstr";

	if (signal(SIGINT, SIG_IGN) == SIG_DFL)
		(void)signal(SIGINT, onintr);
	if (cflg || (argc == 0 && !readstd))
		inithash();
	else {
		int	fd;

		snprintf(stringtmpfile, sizeof(stringtmpfile),
		    "%s%s.XXXXXX", _PATH_TMP, "xstr");
		strings = stringtmpfile;
		fd = mkstemp(stringtmpfile);
		if (fd == -1)
			err(1, "mkstemp failed");
		close(fd);
	}
	while (readstd || argc > 0) {
		if (freopen("x.c", "w", stdout) == NULL)
			err(1, "Cannot open `%s'", "x.c");
		if (!readstd && freopen(argv[0], "r", stdin) == NULL)
			err(1, "Cannot open `%s'", argv[0]);
		process("x.c");
		if (readstd == 0)
			argc--, argv++;
		else
			readstd = 0;
	};
	flushsh();
	if (cflg == 0)
		xsdotc();
	if (strings[0] == '/')
		(void)unlink(strings);
	exit(0);
}

static void
process(const char *name)
{
	char *cp;
	int c;
	int incomm = 0;
	int inasm = 0;
	int asmparnest = 0;
	int ret;

	printf("extern char\t%s[];\n", array);
	for (;;) {
		if (fgets(linebuf, sizeof linebuf, stdin) == NULL) {
			if (ferror(stdin))
				err(1, "Error reading `%s'", name);
			break;
		}
		if (linebuf[0] == '#') {
			printf("%s", linebuf);
			continue;
		}
		for (cp = linebuf; (c = *cp++);)
			switch (c) {

			case '"':
				if (incomm || inasm)
					goto def;
				if ((ret = (int) yankstr(&cp)) == -1)
					goto out;
				printf("(&%s[%d])", array, ret);
				break;

			case '\'':
				if (incomm || inasm)
					goto def;
				putchar(c);
				if (*cp)
					putchar(*cp++);
				break;

			case '/':
				if (incomm || *cp != '*')
					goto def;
				incomm = 1;
				cp++;
				printf("/*");
				continue;

			case '*':
				if (incomm && *cp == '/') {
					incomm = 0;
					cp++;
					printf("*/");
					continue;
				}
				goto def;

			case '(':
				if (!incomm && inasm)
					asmparnest++;
				goto def;

			case ')':
				if (!incomm && inasm && !--asmparnest)
					inasm = 0;
				goto def;

			case '_':
				if (incomm || inasm)
					goto def;
				if (!strncmp(cp, "_asm", 4)) {
					cp += 4;
					printf("__asm");
					if (!strncmp(cp, "__", 2)) {
						cp += 2;
						printf("__");
					}
					if (isalnum((unsigned char)*cp) ||
					    *cp == '_')
						goto def;
					asmparnest = 0;
					inasm = 1;
				} else
					goto def;
				break;
def:
			default:
				putchar(c);
				break;
			}
	}
out:
	if (ferror(stdout)) {
		warn("Error reading `%s'", "x.c");
		onintr(1);
	}
}

static off_t
yankstr(char **cpp)
{
	char *cp = *cpp;
	int c, ch;
	char *dbuf, *dp, *edp;
	const char *tp;
	off_t hash;
	size_t bsiz = BUFSIZ;

	if ((dp = dbuf = malloc(bsiz)) == NULL)
		err(1, "malloc");
	edp = dbuf + bsiz;

	while ((c = *cp++) != '\0') {
		switch (c) {

		case '"':
			/* Look for a concatenated string */
			for (;;) {
				while (isspace((unsigned char)*cp))
					cp++;
				if (*cp == '\0') {
					if (fgets(linebuf,
					    sizeof linebuf, stdin) == NULL) {
						if (ferror(stdin))
							err(1,
							"Error reading `x.c'");
						goto out;
					}
					cp = linebuf;
				} else {
					if (*cp == '"') {
						cp++;
						if (*cp == '"') {
							cp++;
							continue;
						} else {
							c = *cp++;
							goto gotc;
						}
					} else {
						cp++;
						goto out;
					}
				}
			}
			/*NOTREACHED*/
		case '\\':
			c = *cp++;
			if (c == 0)
				break;
			if (c == '\n') {
				if (fgets(linebuf, sizeof linebuf, stdin)
				    == NULL) {
					if (ferror(stdin))
						err(1, "Error reading `x.c'");
					return(-1);
				}
				cp = linebuf;
				continue;
			}
			for (tp = "b\bt\tr\rn\nf\f\\\\\"\""; (ch = *tp++); tp++)
				if (c == ch) {
					c = *tp;
					goto gotc;
				}
			if (!octdigit(c)) {
				*dp++ = '\\';
				break;
			}
			c -= '0';
			if (!octdigit(*cp))
				break;
			c <<= 3, c += *cp++ - '0';
			if (!octdigit(*cp))
				break;
			c <<= 3, c += *cp++ - '0';
			break;
		}
gotc:
		if (dp >= edp - 1) {
			char *nbuf;
			bsiz += BUFSIZ;
			if ((nbuf = realloc(dbuf, bsiz)) == NULL) {
				free(dbuf);
				err(1, "realloc");
			}
			dp = nbuf + (dp - dbuf);
			edp = nbuf + bsiz;
			dbuf = nbuf;
		}
		*dp++ = c;
	}
out:
	*cpp = --cp;
	*dp = '\0';
	hash = hashit(dbuf, 1);
	free(dbuf);
	return hash;
}

static int
octdigit(char c)
{

	return (isdigit((unsigned char)c) && c != '8' && c != '9');
}

static void
inithash(void)
{
	char buf[BUFSIZ];
	FILE *mesgread = fopen(strings, "r");

	if (mesgread == NULL)
		return;
	for (;;) {
		mesgpt = tellpt;
		if (fgetNUL(buf, sizeof buf, mesgread) == 0)
			break;
		(void)hashit(buf, 0);
	}
	(void)fclose(mesgread);
}

static int
fgetNUL(char *obuf, int rmdr, FILE *file)
{
	int c;
	char *buf = obuf;

	c = 0;	/* XXXGCC -Wuninitialized */

	while (--rmdr > 0 && (c = xgetc(file)) != 0 && c != EOF)
		*buf++ = c;
	*buf++ = 0;
	return (feof(file) || ferror(file)) ? 0 : 1;
}

static int
xgetc(FILE *file)
{

	tellpt++;
	return getc(file);
}


static off_t
hashit(const char *str, int new)
{
	int i;
	struct hash *hp, *hp0;

	hp = hp0 = &bucket[lastchr(str) & 0177];
	while (hp->hnext) {
		hp = hp->hnext;
		i = istail(str, hp->hstr);
		if (i >= 0)
			return (hp->hpt + i);
	}
	if ((hp = calloc(1, sizeof (*hp))) == NULL)
		err(1, NULL);
	hp->hpt = mesgpt;
	if ((hp->hstr = strdup(str)) == NULL)
		err(1, NULL);
	mesgpt += strlen(hp->hstr) + 1;
	hp->hnext = hp0->hnext;
	hp->hnew = new;
	hp0->hnext = hp;
	return (hp->hpt);
}

static void
flushsh(void)
{
	int i;
	struct hash *hp;
	FILE *mesgwrit;
	int old = 0, new = 0;

	for (i = 0; i < BUCKETS; i++)
		for (hp = bucket[i].hnext; hp != NULL; hp = hp->hnext)
			if (hp->hnew)
				new++;
			else
				old++;
	if (new == 0 && old != 0)
		return;
	mesgwrit = fopen(strings, old ? "r+" : "w");
	if (mesgwrit == NULL)
		err(1, "Cannot open `%s'", strings);
	for (i = 0; i < BUCKETS; i++)
		for (hp = bucket[i].hnext; hp != NULL; hp = hp->hnext) {
			found(hp->hnew, hp->hpt, hp->hstr);
			if (hp->hnew) {
				(void)fseek(mesgwrit, hp->hpt, 0);
				(void)fwrite(hp->hstr, strlen(hp->hstr) + 1, 1,
				    mesgwrit);
				if (ferror(mesgwrit))
					err(1, "Error writing `%s'", strings);
			}
		}
	if (fclose(mesgwrit) == EOF)
		err(1, "Error closing `%s'", strings);
}

static void
found(int new, off_t off, const char *str)
{
	if (vflg == 0)
		return;
	if (!new)
		(void)fprintf(stderr, "found at %d:", (int) off);
	else
		(void)fprintf(stderr, "new at %d:", (int) off);
	prstr(str);
	(void)fprintf(stderr, "\n");
}

static void
prstr(const char *cp)
{
	int c;

	while ((c = (*cp++ & 0377)) != '\0')
		if (c < ' ')
			(void)fprintf(stderr, "^%c", c + '`');
		else if (c == 0177)
			(void)fprintf(stderr, "^?");
		else if (c > 0200)
			(void)fprintf(stderr, "\\%03o", c);
		else
			(void)fprintf(stderr, "%c", c);
}

static void
xsdotc(void)
{
	FILE *strf = fopen(strings, "r");
	FILE *xdotcf;

	if (strf == NULL)
		err(1, "Cannot open `%s'", strings);
	xdotcf = fopen("xs.c", "w");
	if (xdotcf == NULL)
		err(1, "Cannot open `%s'", "xs.c");
	(void)fprintf(xdotcf, "char\t%s[] = {\n", array);
	for (;;) {
		int i, c;

		for (i = 0; i < 8; i++) {
			c = getc(strf);
			if (ferror(strf)) {
				warn("Error reading `%s'", strings);
				onintr(1);
			}
			if (feof(strf)) {
				(void)fprintf(xdotcf, "\n");
				goto out;
			}
			(void)fprintf(xdotcf, "0x%02x,", c);
		}
		(void)fprintf(xdotcf, "\n");
	}
out:
	(void)fprintf(xdotcf, "};\n");
	(void)fclose(xdotcf);
	(void)fclose(strf);
}

static char
lastchr(const char *cp)
{

	while (cp[0] && cp[1])
		cp++;
	return (*cp);
}

static int
istail(const char *str, const char *of)
{
	int d = strlen(of) - strlen(str);

	if (d < 0 || strcmp(&of[d], str) != 0)
		return (-1);
	return (d);
}

static void
onintr(int dummy)
{

	(void)signal(SIGINT, SIG_IGN);
	if (strings[0] == '/')
		(void)unlink(strings);
	(void)unlink("x.c");
	(void)unlink("xs.c");
	exit(dummy);
}

static void
usage(void)
{

	(void)fprintf(stderr, "usage: %s [-cv] [-l array] [-] [<name> ...]\n",
	    getprogname());
	exit(1);
}
